### PR TITLE
🪟🐛 client side validation of cron expressions

### PR DESCRIPTION
**Pull request by joeykmh**
## What
Closes https://github.com/airbytehq/airbyte-cloud/issues/2772

Adds frontend validation of cron expressions

## How
I tested various cron parsing libraries, but unfortunately did not find any that work perfectly with the [quartz syntax](http://www.quartz-scheduler.org/documentation/quartz-2.3.0/tutorials/crontrigger.html).

Instead of parsing the cron expression, I'm relying on a pretty loose regex for the frontend validation. This will give the user quick feedback if, for example, the cron is too short or has invalid characters. I have included tests for all [examples on the quartz docs](https://www.quartz-scheduler.org/documentation/quartz-2.3.0/tutorials/crontrigger.html#examples) as well as every single cron expression one of our cloud customers is currently using.

It is possible that an invalid cron expression makes it through the frontend validation. In that case, the backend will return an error which we display to the user.

### CODE REVIEW

File: airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx [hunk 1]
```diff
@@ -99,7 +100,12 @@ export const connectionValidationSchema = (mode: ConnectionFormMode) =>
         return yup.object({
           cron: yup
             .object({
-              cronExpression: yup.string().required("form.empty.error"),
+              cronExpression: yup
+                .string()
+                .required("form.empty.error")
+                .test("validCron", "form.cronExpression.error", (expression) =>
+                  !expression ? false : validateCronExpression(expression)
```
[the remark this answers is not in the document]
Comment (josephkmh) [reply airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx:107]: Yeah that's probably a good call. `Yup.string.required()` unfortunately doesn't narrow the type (it's a [bug](https://github.com/jquense/yup/issues/1367)), so it's still `string | undefined` when it reaches the `.test()` method.

File: airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/validateCronExpression.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  try {
+    const cronFields = expression.trim().split(" ");
+
+    if (cronFields.length < 6) {
+      throw new Error(
+        `Cron expression "${expression}" must contain at least 6 fields (${cronFields.length} fields found)`
+      );
+    }
+
+    if (cronFields.length > 7) {
```
[the remark this answers is not in the document]
Comment (josephkmh) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/validateCronExpression.ts:21]: Yes that's an optional field in the [quartz cron syntax](http://www.quartz-scheduler.org/documentation/quartz-2.3.0/tutorials/crontrigger.html).

File: airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx [hunk 3]
```diff
@@ -29,6 +29,7 @@ import { ValuesProps } from "hooks/services/useConnectionHook";
 import { useCurrentWorkspace } from "services/workspaces/WorkspacesService";
 
 import calculateInitialCatalog from "./calculateInitialCatalog";
+import { validateCronExpression } from "./ScheduleField";
```
[the remark this answers is not in the document]
Comment (josephkmh) [reply airbyte-webapp/src/views/Connection/ConnectionForm/formConfig.tsx:32]: Sure, I can move it there. I wonder if over time `src/utils` might get bloated if we move all pure/generic functions there, but we can cross that bridge when we get there.

File: airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx [hunk 4]
```diff
@@ -204,6 +203,11 @@ const ScheduleField: React.FC = () => {
                     onChange={(item: DropDownOptionDataItem) => onCronChange(item, field, form, "cronTimeZone")}
                   />
                 </div>
+                {(form.errors?.scheduleData as ConnectionScheduleData)?.cron?.cronExpression && (
```
[the remark this answers is not in the document]
Comment (josephkmh) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx:206]: Yeah it's kind of a nasty expression, I'll extract it to make the JSX cleaner.

File: airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx [hunk 5]
```diff
@@ -21,13 +23,12 @@ const CRON_DEFAULT_VALUE = {
   cronExpression: "0 0 12 * * ?",
 };
 
-const CRON_REFERENCE_LINK = "http://www.quartz-scheduler.org/documentation/quartz-2.3.0/tutorials/crontrigger.html";
```
Comment (timroes) [inline airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx:24]: :bow:

File: airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx [hunk 6]
```diff
@@ -128,6 +129,10 @@ const ScheduleField: React.FC = () => {
     return form.values.scheduleType === ConnectionScheduleType.cron;
   };
 
+  const cronValidationError = useMemo(() => {
+    return (errors?.scheduleData as ConnectionScheduleData)?.cron?.cronExpression;
+  }, [errors]);
```
Comment (timroes) [inline airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx:134]: If we want to get rid of the typecasting here (which imho looks a bit weird, given that TS thinks `scheduleData` is `string` here, we could directly pull the error of that field (and get rid of the whole `useMemo` here):

```ts
  const [, { error }] = useField("scheduleData.cron.cronExpression");
```

**Review by timroes — APPROVED** [review]
Have tested locally. Works for required and invalid cron expressions. Code LGTM, I left one suggestion to get rid of a typecast.

File: airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx (hunk 6 above)
Comment (josephkmh) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ScheduleField/ScheduleField.tsx:134]: Nice, that is cleaner, I'll use that. It still is not super type safe (`useField()` will accept any string without complaining) but it's not as ugly 😅 

I think to get proper typing we would need to change the yup schema away from overloading `scheduleData` with the ambiguous `.mixed().when(...)`, but I didn't want to blow up the PR.

**Pull request merged**